app/models/grda_warehouse/ch_enrollment.rb

```diff
@@ -74,5 +74,324 @@ def self.update_existing!
+        return steps if steps.last[:result]
+      end
+
+      # This matches the last step (26) for clients who were homeless 4-11 months, since 'homeless_duration_sufficient' doesn't check for that
+      steps.last[:result] = :no
+      steps
+    end
+
+    def self.dk_or_r_or_missing(value)
+      return :dk_or_r if [8, 9].include?(value)
+      return :missing if [nil, 99].include?(value)
+    end
+
+    def self.is_no?(value) # rubocop:disable Naming/PredicateName
+      return :no if value&.zero?
+    end
+
+    # Line 1 (3.08)
+    def self.disabling_condition(enrollment)
+      result = if is_no?(enrollment.DisablingCondition)
+        :no
+      elsif dk_or_r_or_missing(enrollment.DisablingCondition)
+        dk_or_r_or_missing(enrollment.DisablingCondition)
+      end
+
+      { result: result, display_value: enrollment.DisablingCondition, line: 1 }
+    end
+
+    # Line 3 (2.02.6)
+    def self.project_type(enrollment)
+      ptype = enrollment.project.computed_project_type
+      result = GrdaWarehouse::Hud::Project::CHRONIC_PROJECT_TYPES.include?(ptype)
+      { result: result, display_value: "#{ptype} (#{::HUD.project_type_brief(ptype)})", line: 3 }
+    end
+
+    # Line 9  (3.917.1)
+    def self.prior_living_sitation_homeless(enrollment)
+      value = enrollment.LivingSituation
+      result = HUD.homeless_situations(as: :prior).include?(value)
+      { result: result, display_value: "#{value} (#{::HUD.living_situation(value)})", line: 9 }
+    end
+
+    # Line 14 (3.917.1)
+    def self.prior_living_sitation_institutional(enrollment)
+      value = enrollment.LivingSituation
+      result = HUD.institutional_situations(as: :prior).include?(value)
+      { result: result, display_value: "#{value} (#{::HUD.living_situation(value)})", line: 14 }
+    end
+
+    # Line 21 (3.917.1)
+    def self.prior_living_sitation_other(enrollment)
+      value = enrollment.LivingSituation
+      result = (HUD.temporary_and_permanent_housing_situations(as: :prior) + HUD.other_situations(as: :prior)).include?(value)
+      { result: result, display_value: "#{value} (#{::HUD.living_situation(value)})", line: 21 }
+    end
+
+    # Lines 4, 10, 17, and 24 (3.917.3)
+    def self.approximate_start_date(enrollment, date: enrollment.EntryDate)
+      ch_start_date = [enrollment.DateToStreetESSH, enrollment.EntryDate].compact.min
+      project = enrollment.project
+      days = if date != enrollment.EntryDate && (project.so? || project.es? && project.bed_night_tracking?)
+        dates_in_enrollment_between(enrollment, enrollment.EntryDate, date).count + (enrollment.EntryDate - ch_start_date).to_i
+      else
+        (date - ch_start_date).to_i
+      end
+      result = days > 365 ? :yes : nil
+      { result: result, display_value: "#{days} days" }
+    end
+
+    THREE_OR_FEWER_TIMES_HOMELESS = [1, 2, 3].freeze
+    TWELVE_OR_MORE_MONTHS_HOMELESS = [112, 113].freeze # 112 = 12 months, 113 = 13+ months
```

Comment:
We generally put constants at the top of the class so we know they exist.  Can you move them up?

Reply:
Alternatively, since these are only used once, and they are very small arrays, I'm not sure there's much benefit to having them defined as constants.

Reply:
Okay, I was just trying to mirror what we had before re- memoization. I think the code is easier to read if the values are next to where they are used (to understand `- 100` bit afterwards) so I will just move them back into the functions.